src/lib/operators/table_scan.cpp

```diff
@@ -0,0 +1,295 @@
+   * 2. Check for the type of the Segment
+   * 3. Specific for each type compare each value in the segment against the compare value with the given comparator
+   *    to create position_list
+   */
+
+  const auto num_chunks = input_table->chunk_count();
+  std::vector<RowID> position_list;
+
+  resolve_data_type(input_table->column_type(_column_id), [&](auto type) {
+    using Type = typename decltype(type)::type;
+    Type typed_search_value = type_cast<Type>(_search_value);
+
+    Assert((_search_value.type() == typeid(Type)), "Type mismatch of the type of the search value and the column type");
+
+    // Iterate over each Chunk / the relevant Segment per Chunk referenced by the given ColumnID
+    for (auto chunk_id = ChunkID{0}; chunk_id < num_chunks; ++chunk_id) {
+      auto segment = input_table->get_chunk(chunk_id).get_segment(_column_id);
+
+      // Check for segment type
+      const auto typed_value_segment = std::dynamic_pointer_cast<ValueSegment<Type>>(segment);
+
+      if (typed_value_segment != nullptr) {
+        // For ValueSegments simply compare each value of the value segment
+        // against the compare value with given comparator
+        std::vector<Type> values = typed_value_segment->values();
+        auto comparator = _get_comparator<Type>(_scan_type);
+
+        for (auto cell_id = 0ul, typed_segment_size = values.size(); cell_id < typed_segment_size; ++cell_id) {
+          if (comparator(values[cell_id], typed_search_value)) {
+            position_list.push_back(RowID{static_cast<ChunkID>(chunk_id), static_cast<ChunkOffset>(cell_id)});
+          }
+        }
+      } else if (const auto typed_dictionary_segment = std::dynamic_pointer_cast<DictionarySegment<Type>>(segment);
+                 typed_dictionary_segment != nullptr) {
+        /**
+         * For Dictionary Segments use index-based comparing with value ids.
+         * This is possible because dictionary values are stored in sorted manner and the total order of dictionary value ids 
+         * therefore is based on the total order of values in the attribute vector.
+         * Specifically, value_id_a <= value_id_b iff. value_a <= value_b.
+         * We compare by first retrieving the correct compare_value_id to compare against given the ScanType and search value, 
+         * and then comparing the ValueIDs in the attribute vector against it.
+         */
+
+        auto dictionary_comparator = _get_comparator<ValueID>(_scan_type);
+        auto attribute_value_ids = typed_dictionary_segment->attribute_vector();
+
+        const auto search_value_id =
+            _get_compare_value_id(_scan_type, typed_dictionary_segment->upper_bound(typed_search_value),
+                                  typed_dictionary_segment->lower_bound(typed_search_value));
+
+        if (search_value_id == EMPTY_RESULT) {
+          continue;
+        }
+
+        if (search_value_id == SELECT_EVERYTHING) {
```

Comment:
Could it happen that `_get_compare_value_id` returns an actual ValueID for a large full 32bit dictionary that matches `SELECT_EVERYTHING` or `EMPTY_RESULT`, i.e., max of ValueID minus 1 or 2? In that case you would produce wrong results, wouldn't you?

Reply:
If this is somehow handled differently and still correct, it is too hard to understand and encoding such information in const ValueIDs does not seem to be the best way to go.